pysb/simulator/scipyode.py

```diff
@@ -551,6 +417,353 @@ def _integrator_process(initials, param_values, code_eqs, jac_eqs, num_species,
+        Whether to delete the work directory with generated code and compiled
+        output files upon destruction.
+
+    Attributes
+    ----------
+    with_jacobian : bool
+        Whether to construct the Jacobian evaluation function.
+    cleanup : bool
+        Whether to delete the work directory with generated code and compiled
+        output files upon destruction.
+    y : sympy.MatrixSymbol
+        Symbol "y" representing the amount of all model species.
+    p : sympy.MatrixSymbol
+        Symbol "p" representing all model parameters.
+    e : sympy.MatrixSymbol
+        Symbol "e" representing all model constant expressions.
+    o : sympy.MatrixSymbol
+        Symbol "o" representing all model observables.
+    kinetics : sympy.Matrix
+        Symbolic form of model reaction rates in terms of y, p, e, o, and time.
+    kinetics_jacobian_y : sympy.SparseMatrix
+        Symbolic form of partial derivatives of kinetics with respect to y
+        (Only set if with_jacobian is True)
+    kinetics_jacobian_o : sympy.SparseMatrix
+        Symbolic form of partial derivatives of kinetics with respect to o
+        (Only set if with_jacobian is True)
+    stoichiometry_matrix : scipy.sparse.csr_matrix
+        The model's stoichiometry matrix.
+    observables_matrix : scipy.sparse.csr_matrix
+        Encodes the linear combinations of species that define the model's
+        observables.
+    model_name : str
+        The name of the model.
+
+    """
+
+    def __init__(self, model, with_jacobian=False, cleanup=True):
+        self.with_jacobian = with_jacobian
+        self.cleanup = cleanup
+        expr_dynamic = model.expressions_dynamic(include_derived=True)
+        expr_constant = model.expressions_constant(include_derived=True)
+        self.y = sympy.MatrixSymbol('y', len(model.species), 1)
+        self.p = sympy.MatrixSymbol('p', len(model.parameters_all()), 1)
+        self.e = sympy.MatrixSymbol('e', len(expr_constant), 1)
+        self.o = sympy.MatrixSymbol('o', len(model.observables), 1)
+        species_subs = {
+            sympy.Symbol('__s%d' % i): self.y[i]
+            for i, _ in enumerate(model.species)
+        }
+        param_subs = dict(zip(model.parameters_all(), self.p))
+        obs_subs = dict(zip(model.observables, self.o))
+        expr_dynamic_subs = {
+            e: e.expand_expr()
+            for i, e in enumerate(expr_dynamic)
+        }
+        expr_constant_subs = dict(zip(expr_constant, self.e))
+        obs_matrix = scipy.sparse.lil_matrix(
+            (len(model.observables), len(model.species)), dtype=np.int64
+        )
+        for i, obs in enumerate(model.observables):
+            obs_matrix[i, obs.species] = obs.coefficients
+        self.kinetics = sympy.Matrix([
+            r['rate']
+            .xreplace(expr_constant_subs)
+            .xreplace(expr_dynamic_subs)
+            .xreplace(param_subs)
+            .xreplace(species_subs)
+            .xreplace(obs_subs)
```

Comment:
Would it be more efficient here to store the various subs as a generator (e.g. using `itertools.chain`) to allow a single `xreplace` call? This would obviate the need for multiple loops over the symbols in `r['rate']`, but maybe it wouldn't be a significant speedup. Just a thought.

Reply:
The performance isn't an issue but I'll rework it to only call `xreplace` once per reaction anyway. `xreplace` takes a dict so `chain` wouldn't work, but I can just merge all the dicts and use that. I need to preserve entry order though -- should I use an OrderedDict or rely on the official dict ordering in python 3.7+? What python releases will we want to support?

Reply:
Oops, `xreplace` doesn't handle the case where the result of one substitution needs to also be processed by a later one. `subs` does handle this, but I thought the performance was worse. I'll benchmark before committing to an approach.